torch/_refs/__init__.py

```diff
@@ -3729,62 +3729,111 @@ def repeat(a: Tensor, *repeat_shape) -> Tensor:
     return permuted_result.reshape(target_shape)
 
 
-def _reshape_view_helper(a: TensorLikeType, *shape, allow_copy: bool) -> TensorLikeType:
-    from torch._dynamo.exc import UserError, UserErrorType
+# this function is python match of computeStride_impl in TensorUtils.cpp
+def _compute_stride(old_shape, old_stride, new_shape):
     from torch.fx.experimental.symbolic_shapes import (
         guard_or_false,
         guard_or_true,
-        GuardOnDataDependentSymNode,
+        sym_eq,
     )
 
+    if len(old_shape) == 0:
+        return [1] * len(new_shape)
+
+    numel = reduce(operator.mul, old_shape, 1)
+    zero_numel = guard_or_false(numel == 0)
+    if zero_numel and guard_or_false(sym_eq(old_shape, new_shape)):
+        return old_stride
+
+    new_stride = [0] * len(new_shape)
+
+    if zero_numel:
+        for view_d in range(len(new_shape) - 1, -1, -1):
+            if view_d == len(new_shape) - 1:
+                new_stride[view_d] = 1
+            else:
+                new_stride[view_d] = (
+                    max(new_shape[view_d + 1], 1) * new_stride[view_d + 1]
+                )
+        return new_stride
+
+    view_d = len(new_shape) - 1
+    chunk_base_stride = old_stride[-1]
+    tensor_numel = 1
+    view_numel = 1
+
+    for tensor_d in range(len(old_shape) - 1, -1, -1):
+        tensor_numel *= old_shape[tensor_d]
+
+        if tensor_d == 0 or (
+            guard_or_true(old_shape[tensor_d - 1] != 1)
+            and guard_or_true(
+                old_stride[tensor_d - 1] != tensor_numel * chunk_base_stride
+            )
+        ):
+            while view_d >= 0 and (
+                guard_or_true(view_numel < tensor_numel)
+                or guard_or_false(new_shape[view_d] == 1)
+            ):
+                new_stride[view_d] = view_numel * chunk_base_stride
+                view_numel *= new_shape[view_d]
+                view_d -= 1
+
+            if guard_or_true(view_numel != tensor_numel):
+                return None
+
+            if tensor_d > 0:
+                chunk_base_stride = old_stride[tensor_d - 1]
+                tensor_numel = 1
+                view_numel = 1
+    if view_d != -1:
+        return None
+    return new_stride
+
+
+# if a is contiguous, we can always reshape with as_strided and contiguous_strides.
+# if a is not contiguous and _compute_stride succeed we also use as_strided without clone.
```

Comment:
Can you please elaborate on these cases? For these correctness comments, you don't need to be frugal with your words. Most of the comments sound like notes for yourself, but for these comments you should really imagine what another core pytorch dev reading this comment 6 months from now will think.